exercises/nondet_monad_defun/descr.md

```diff
@@ -0,0 +1,530 @@
+contain `tick`s, the automated grading system can tell when
+computations are executed. For instance, demanding just the first
+result of the computation
+`choose (tick (return 0)) (tick (return 1))` should cause `work` to
+be incremented just once, not twice.
+
+*Note.* The automated grading system for Question 1 does not exercise the
+cases for `FSplit`, `SSplit` and `MReflect` in the abstract machine. In order
+to ensure that your code deals with these cases correctly, you must continue
+this exercise and pass Questions 2-4 as well.
+
+As in the continuation-based implementation described in a
+<a href="" onclick="top.location='/exercises/nondet_monad_cont/';">companion exercise</a>,
+before attacking Questions 2-4,
+we must define an auxiliary function, `msplit`.
+
+The need for `msplit` arises from the following remark. When executed, a
+computation of type `'a m` produces a sequence of results. Such a sequence
+must be either empty or nonempty, and in the latter case, it must have a first
+element and a remainder. If a computation *was* just a sequence (that is, if
+the type `'a m` was a synonym for `'a Seq.t`), then we could distinguish these
+two situations just by performing a case analysis.
+
+Because we have defined the type `'a m` in a different way, a direct case
+analysis is impossible. Instead, the idea is to define a function `msplit`
+that *converts* from the type `'a m` to the type `unit -> ('a * 'a m) option`.
+This opens the door to a case analysis. In short, `msplit` allows us to
+perform case analysis on a computation *as if* it was a sequence, even though
+its internal representation is different.
+
+**Question X.** (Not graded.)
+Complete the implementation of the auxiliary function `msplit.`
+
+Once `msplit` is available,
+Questions 2-4 are answered in exactly the same way as in the
+<a href="" onclick="top.location='/exercises/nondet_monad_cont/';">companion exercise</a>.
+
+**Question 2.** Implement the constructor function `at_most_once`.
+
+*Hint.* Use `msplit` to determine if the computation at hand fails or produces
+at least one result. In the first case, recall that `at_most_once fail` is
+supposed to behave like `fail`. In the second case, recall that
+`at_most_once (choose (return x) m)` is supposed to behave like
+`return x`.
+
+*Hint.* Use `delay` to ensure that the case analysis is performed when
+the computation `at_most_once m` is *executed*, not when it is *built*.
+
+**Question 3.** Implement the constructor function `interleave`.
+
+*Hint.* Again, use `delay` and `msplit` in an appropriate manner.
+Then, ask yourself how `interleave fail m2` is supposed to behave,
+and how `interleave (choose (return x1) m1) m2` is supposed
+to behave.
+
+**Question 4.** Implement the constructor function `(>>-)`.
+
+*Hint.* Again, use `delay` and `msplit` in an appropriate manner.
+Then, ask yourself how `fail >>- m2` is supposed to behave,
+and how `(choose (return x1) m1) >>- m2` is supposed
+to behave.
+
+## Notes
```

Comment:
It is a bit frustrating not to have a comparison between the two implementations. Yet, I guess that comparison is not that easy to conduct.

Reply:
Which two implementations would you like to compare?

Reply:
I was referring to the defunctionalized approach and the continuation-based approach.

Reply:
Would you want to compare them in terms of performance or in terms of architecture? Regarding performance, I don't think this would be very interesting. Regarding architecture, it is indeed interesting to compare them. I think that if a student does both exercises, then they can understand that they both have the same underlying structure. Should I explain more?

Reply:
They both have the same underlying structure but they are still distinct. What makes a functional programmer choose one approach over the other to solve this problem? Is it just a matter of style? I guess we could at least say that the defunctionalized approach can be implemented in a non-functional language. Is there more to say? I do not know.

Reply:
I have added the comment that you suggest. Other than that, I have already written that I think the functional version is easier to discover than the defunctionalized approach. I don't know if there is anything more to say...